### PR TITLE
Create API utility base - Closes #3161

### LINKED ISSUE #3161: Create API utility base
### Description
To kick start the refactoring of the API utility methods as explained in #2890, we need to handle the followings first:
- JavaScript module structure
- A mapper function to call the right function according to the active token
- A generic function to standardize the HTTP calls, response and error handling.
- A generic function to standardize JSON RPC calls, response and error handling.

The goal is for the rest of the tickets in #2890 to use the above generic functions for easily and consistently handling API calls.

### Acceptance criteria
- Sufficient unit test coverage
- Proper function signature documentation

### PR DESCRIPTION
**Pull request by reyraa**
### What was the problem?
This PR resolves #3161

### CODE REVIEW

File: src/store/middlewares/network.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ const getServerUrl = (nodeUrl, nethash) => {
   if (/\.(liskdev.net|lisk.io)$/.test(nodeUrl)) {
     return nodeUrl.replace(/\.(liskdev.net|lisk.io)$/, $1 => `-service${$1}`);
   }
-  return 'unavailable';
+  return 'http://localhost:9091';
```
Comment (UsamaHameed) [inline src/store/middlewares/network.js:21]: is this a permanent change or should be removed later?

File: src/utils/api/http.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * if fails on server it throws an error,
+ *
+ */
+
+const http = ({
+  baseUrl, path, params, method, network, ...restOptions
+}) => {
+  const url = baseUrl ? `${baseUrl}${path}` : `${network.serviceUrl}${path}`;
+  return fetch(url, {
+    method,
```
Comment (UsamaHameed) [inline src/utils/api/http.js:20]: can we default the method to "GET"?

File: src/utils/api/http.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    mode: 'no-cors',
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify(params),
```
Comment (UsamaHameed) [inline src/utils/api/http.js:25]: The params for a get request would have to be a part of the url so we need to rethink this

File: src/utils/api/http.test.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import http from './http';
+
+describe('HTTP', () => {
+  const data = {
+    path: '/api/enpoint/',
```
Comment (UsamaHameed) [inline src/utils/api/http.test.js:5]:
```suggestion
    path: '/api/endpoint/',
```

File: src/utils/api/http.test.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  it('should be able to set an optional base url', async () => {
+    const baseUrl = 'http://testnet.net';
+    global.fetch = jest.fn(() => Promise.resolve({
+      ok: true,
+      status: 200,
+      json: () => Promise.resolve('ok'),
+    }));
+    await http(data);
+    await http({ ...data, baseUrl });
+    expect(fetch.mock.calls[0][0]).toEqual(`${data.network.serviceUrl}${data.path}`);
```
Comment (UsamaHameed) [inline src/utils/api/http.test.js:20]: We should use `expect(fetch).toHaveBeenCalledWith` instead

File: src/utils/api/http.test.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  });
+
+  it('should return data', async () => {
+    const expectedResponse = { test: ['ok'] };
+    global.fetch = jest.fn(() => Promise.resolve({
+      ok: true,
+      status: 200,
+      json: () => Promise.resolve(expectedResponse),
+    }));
+    await expect(http(data)).resolves.toEqual(expectedResponse);
```
Comment (UsamaHameed) [inline src/utils/api/http.test.js:33]: Why not this? 
```
const response = await http(data);
...
// assert on response? 

```

File: src/utils/api/http.test.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  it('should throw error', async () => {
+    const statusText = 'Error meanwhile processing the call';
+    global.fetch = jest.fn(() => Promise.resolve({
+      ok: false,
+      status: 500,
+      statusText,
+      json: () => Promise.resolve('test'),
+    }));
+    await expect(http(data)).rejects.toEqual(Error(statusText));
```
Comment (UsamaHameed) [inline src/utils/api/http.test.js:44]: Same as above

File: src/utils/api/ws.test.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    const emit = jest.fn().mockImplementation((evtName, params, callback) => {
+      callback({ result: {} });
+    });
+    io.mockImplementation(() => ({ emit }));
+
+    const wsPromise = ws({
+      baseUrl,
+      requests,
+    });
+    await wsPromise;
```
Comment (UsamaHameed) [inline src/utils/api/ws.test.js:23]: We can await in the line above

File: src/utils/api/ws.test.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    expect(typeof wsPromise.then).toEqual('function');
+    expect(typeof wsPromise.catch).toEqual('function');
```
Comment (UsamaHameed) [inline src/utils/api/ws.test.js:26]: No need to assert for these since `then` and `catch` are always functions 🤔

File: src/utils/api/ws.test.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    expect(io.mock.calls[0][0]).toContain(baseUrl);
+    expect(emit).toHaveBeenCalledTimes(1);
+    expect(emit.mock.calls[0][0]).toEqual('request');
+    expect(emit.mock.calls[0][1]).toEqual(requests);
+    expect(typeof emit.mock.calls[0][2]).toEqual('function');
```
Comment (UsamaHameed) [inline src/utils/api/ws.test.js:31]: Prefer the jest wrapper for mocked functions wherever we can.

**Review by UsamaHameed — CHANGES_REQUESTED** [review]

File: src/store/middlewares/network.js (hunk 1 above)
Comment (reyraa) [reply src/store/middlewares/network.js:21]: We should figure out a better way. For now, this saves time.

File: src/utils/api/http.js (hunk 2 above)
Comment (reyraa) [reply src/utils/api/http.js:20]: Done

File: src/utils/api/ws.test.js (hunk 9 above)
Comment (reyraa) [reply src/utils/api/ws.test.js:26]: Initially I wasn't testing the reopens in this test case.

**Review by UsamaHameed — APPROVED** [review]

**Pull request merged**